pkg/sentry/fsimpl/fuse/dev.go

```diff
@@ -46,55 +49,230 @@ type DeviceFD struct {
 	vfs.DentryMetadataFileDescriptionImpl
 	vfs.NoLockFD
 
-	// TODO(gvisor.dev/issue/2987): Add all the data structures needed to enqueue
-	// and deque requests, control synchronization and establish communication
-	// between the FUSE kernel module and the /dev/fuse character device.
+	// mounted specifies whether a FUSE filesystem was mounted using the DeviceFD.
+	mounted bool
+
+	// nextOpID is used to create new requests.
+	nextOpID linux.FUSEOpID
+
+	// queue is the list of requests that need to be processed by the FUSE server.
+	queue requestList
+
+	// completions is used to map a request to its response. A Writer will use this
+	// to notify the caller of a completed response.
+	completions map[linux.FUSEOpID]*FutureResponse
+
+	readCursor  uint32
+	writeCursor uint32
+
+	// writeBuf is the memory buffer used to copy in the FUSE out header from
+	// userspace.
+	writeBuf []byte
+
+	// writeCursorFR current FR being copied from server
+	writeCursorFR *FutureResponse
+
+	// mu protects the communication data structures.
+	mu sync.Mutex
+
+	// waitCh is a channel used to synchronize the readers with the writers.
+	// Readers (FUSE daemon server) block if no requests are available. Writers
+	// (inbound requests to the filesystem) block if there are too many unprocessed
+	// in-flight requests.
+	waitCh chan struct{}
+
 }
 
 // Release implements vfs.FileDescriptionImpl.Release.
 func (fd *DeviceFD) Release() {}
 
 // PRead implements vfs.FileDescriptionImpl.PRead.
 func (fd *DeviceFD) PRead(ctx context.Context, dst usermem.IOSequence, offset int64, opts vfs.ReadOptions) (int64, error) {
+	// Operations on /dev/fuse don't make sense until a FUSE filesystem is mounted.
+	if !fd.mounted {
+		return 0, syserror.EPERM
+	}
+
 	return 0, syserror.ENOSYS
 }
 
 // Read implements vfs.FileDescriptionImpl.Read.
 func (fd *DeviceFD) Read(ctx context.Context, dst usermem.IOSequence, opts vfs.ReadOptions) (int64, error) {
-	return 0, syserror.ENOSYS
+	// Operations on /dev/fuse don't make sense until a FUSE filesystem is mounted.
+	if !fd.mounted {
+		return 0, syserror.EPERM
+	}
+
+	kernelTask := kernel.TaskFromContext(ctx)
+	if kernelTask == nil {
+		log.Warningf("fusefs.DeviceFD.Read: couldn't get kernel task from context")
+		return 0, syserror.EINVAL
+	}
+
+	// Wait for a request to be made available.
+	if err := kernelTask.Block(fd.waitCh); err != nil {
+		log.Warningf("fusefs.DeviceFD.Read: couldn't wait on request queue: %v", err)
+		return 0, syserror.EBUSY
+	}
+
+	fd.mu.Lock()
+	defer fd.mu.Unlock()
+	return fd.readLocked(ctx, dst, opts)
+}
+
+// readLocked
+func (fd *DeviceFD) readLocked(ctx context.Context, dst usermem.IOSequence, opts vfs.ReadOptions) (int64, error) {
+	if fd.queue.Empty() {
+		log.Warningf("fusefs.DeviceFD.Read: No requests to read but still signalled")
+		return 0, syserror.EAGAIN
+	}
+
+	req := fd.queue.Front()
```

Comment:
Is it necessary to sync the queue access using a lock?

Reply:
Yup, so did you update the changes?

Reply:
Ah I see the confusion. This IS protected by a lock is what I meant to say. Everything in `readLocked`  is done while holding a lock. NO changes are required.